src/common/AsyncReserver.h

```diff
@@ -33,24 +35,85 @@ class AsyncReserver {
   unsigned min_priority;
   Mutex lock;
 
-  map<unsigned, list<pair<T, Context*> > > queues;
-  map<T, pair<unsigned, typename list<pair<T, Context*> >::iterator > > queue_pointers;
-  set<T> in_progress;
+  struct Reservation {
+    T item;
+    unsigned prio = 0;
+    Context *grant = 0;
+    Context *preempt = 0;
```

Comment:
Having two Contexts greatly complicates the cleanup: we invoke one of them and need to delete the other. Looking through the code, they *are* being cleaned up by cancel_reservation, but none of the other code paths do so. It's probably best to handle this via a destructor instead, perhaps with some kind of switch we can flip to indicate when we've invoked one or the other?

Reply:
Aha, yeah... We only clean up in preempt_one() and cancel_reservation(); fixing the preempt case to clean up the grant callback!

Reply:
Actually no, it was right before.  The grant context has already been called/used.  I'll zero these pointers so that it is a bit clearer.

Reply:
Ah okay, I forgot that cancel_reservation was both the "undo" and the "I'm done" path, so yes, I think this does work.
...except that means we *do* need to zero them out, or we might delete an object that was reallocated elsewhere!